add blog for how to speed up pod startup from kubelet side

pacoxu opened the pull request:
This blog post discusses methods to speed up pod start-up from the kubelet side, including 
1. enabling parallel image pulls, setting a limit on the number of parallel image pulls at the node level (Limit added in v1.27)
2. increasing default API QPS limits for Kubelet(bump in v1.27)
3. using Evented PLEG. (to beta in v1.27)
4. pod resource limit (most likely the CPU limit will influence it; BTW we changed the throttling factor for cgroup v2 memory, if pod uses a lot of memory during startup, this may be a possible factor)

It also involves the metrics/log of pod startup SLO/SLI.

Other factors that may impact pod startup include container runtime, disk speed, CPU and memory resources on the node.

/cc @ruiwen-zhao @SergeyKanzhelev @harche @wojtek-t

SergeyKanzhelev commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+---
+layout: blog
+title: "How to speed up pod startup from kubelet side? "
```
```suggestion
title: "How to speed up pod startup on a Node."
```

I don't think we need a question mark in a title. Also maybe use Node as a general term.

If this is a feature blog for 1.27, we can title it "Recent developments in kubelet to speed up Pod startup." This will be more accurate as we are not providing other guidances like "minimize the number of config maps Pod uses" or "Do not use too many layers in your image"

SergeyKanzhelev commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+slug: speed-up-pod-startup
+---
+
+**Author**: Paco Xu(DaoCloud), Sergey Kanzhelev, Ruiwen Zhao(Google)
```
```suggestion
**Author**: Paco Xu(DaoCloud), Sergey Kanzhelev (Google), Ruiwen Zhao(Google)
```

SergeyKanzhelev commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+
+How can pod start-up time be accelerated on nodes in large clusters? This is a common issue that cluster administrators may face.
+
+This blog post focuses on methods to speed up pod start-up from the kubelet side. It does not involve the creation time of pods by controller-manager through kube-apiserver, nor does it include scheduling time for pods.
```
```suggestion
This blog post focuses on methods to speed up pod start-up from the kubelet side. It does not involve the creation time of pods by controller-manager through kube-apiserver, nor does it include scheduling time for pods or webhooks executed on it.
```

webhooks is often forgotten and those may delay Pod creation significantly.

SergeyKanzhelev commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+
+## bump default API QPS limits for Kubelet
+
+To improve pod startup in scenarios with multiple pods on a node, particularly sudden scaling situations, it is necessary for Kubelet to synchronize the pod status and prepare configmaps, secrets or volumes. This requires a large bandwidth to access kube-apiserver.
```
```suggestion
To improve pod startup in scenarios with multiple pods on a node, particularly sudden scaling situations, it is necessary for Kubelet to synchronize the pod status and prepare configmaps, secrets, or volumes. This requires a large bandwidth to access kube-apiserver.
```

oxford comma

SergeyKanzhelev commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+
+## Evented PLEG will enhance the scalability of kubelet
+
+`Evented PLEG` is set to be in beta for v1.27, enabled by default. It aids pod startup in large clusters by reducing unnecessary work during inactivity through replacing periodic polling. The current `Generic PLEG` incurs non-negligible overhead due to frequent container status polling, which worsens with Kubelet's parallelism, leading to poor performance and reliability issues. For these reasons ...
```
```suggestion
`Evented PLEG` is set to be in beta for v1.27, disabled by default (needs to be enabled explicitly). It aids pod startup in large clusters by reducing unnecessary work during inactivity through replacing periodic polling. The current `Generic PLEG` incurs non-negligible overhead due to frequent container status polling, which worsens with Kubelet's parallelism, leading to poor performance and reliability issues. For these reasons, it is recommended that we use Evented PLGE instead. Further details can be found in the KEP <https://kep.k8s.io/3386>.
```

ruiwen-zhao commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+
+**Author**: Paco Xu(DaoCloud), Sergey Kanzhelev, Ruiwen Zhao(Google)
+
+How can pod start-up time be accelerated on nodes in large clusters? This is a common issue that cluster administrators may face.
```
nit: this issue is not specific to *large* clusters, right? Even for an one-node cluster, we can use the techniques in this blog to speed up pod startups.

ruiwen-zhao replied:
uh I see. The QPS setting is mostly for large clusters.

pacoxu replied:
Most likely, you will meet the pod start-up thing in a large node or a large cluster.

ruiwen-zhao commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+2. In large clusters they can generate significant load anyway as there are a lot of them
+3. They have a dedicated PriorityLevel and FlowSchema that we can easily control
+
+In old days, we always encounter `volume mount timeout` on kubelet in node with more than 50 pods during pod start up. We suggest our customers to bump 3 or 4 times of kube api QPS and especially with customers that is using bare-metal(physical machines).
```
"3 or 4 times" compared to the old default 5, or the new default 50? Should we make it more clear here?

ruiwen-zhao commented on content/en/blog/_posts/2023-04-02-how-to-speed-up-pod-startup-from-kubelet-side.md:
```diff
@@ -0,0 +1,56 @@
+
+## Evented PLEG will enhance the scalability of kubelet
+
+`Evented PLEG` is set to be in beta for v1.27, enabled by default. It aids pod startup in large clusters by reducing unnecessary work during inactivity through replacing periodic polling. The current `Generic PLEG` incurs non-negligible overhead due to frequent container status polling, which worsens with Kubelet's parallelism, leading to poor performance and reliability issues. For these reasons ...
```
IIUC, Evented PLEG helps reduce startup latency indirectly, by reducing unnecessary works and therefore saving node resources. Cluster size is irrelevant here, right?

pacoxu replied:
Yes, I removed the `large clusters` mentioned here.

ruiwen-zhao commented:
Thanks for the effort! 

/lgtm

sftim commented on content/en/blog/_posts/2023-03-22-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,56 @@
+slug: speed-up-pod-startup
+---
+
+**Author**: Paco Xu(DaoCloud), Sergey Kanzhelev(Google), Ruiwen Zhao(Google)
```
(nit)
```suggestion
**Authors**: Paco Xu (DaoCloud), Sergey Kanzhelev (Google), Ruiwen Zhao (Google)
```

sftim commented on content/en/blog/_posts/2023-03-22-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,56 @@
+---
+layout: blog
+title: "Recent developments in kubelet to speed up Pod startup"
+date: 2023-03-21T16:00:00+0000
```
Is this a post-release blog article? “Recent” suggests these have already shipped, but v1.27 is not released.

pacoxu replied:
Some features mentioned here will be introduced after v1.27. 
I should change this PR to target dev-1.27. Should I change the date to v1.27 release date or keep it as?

pacoxu replied:
It seems that I should update the date here once all content got approved.

sftim commented on content/en/blog/_posts/2023-03-22-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,56 @@
+
+To limit the number of simultaneous image pulls, you can configure the `maxParallelImagePulls` field in kubelet. By setting `maxParallelImagePulls` to a value of _n_, only _n_ images will be pulled concurrently. Any additional image pulls beyond this limit will wait until at least one ongoing pull is complete.
+
+More detials can be found in the KEP <https://kep.k8s.io/3673>.
```
```suggestion
You can find more details in the associated KEP: [Kubelet limit of Parallel Image Pulls](https://kep.k8s.io/3673)
(KEP-3673). The improvement will arrive as part of Kubernetes v1.27 (scheduled for release in April 2023).
```

sftim commented on content/en/blog/_posts/2023-03-22-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,56 @@
+
+More detials can be found in the KEP <https://kep.k8s.io/3673>.
+
+## bump default API QPS limits for Kubelet
```
```suggestion
## Raised default API query-per-second limits for kubelet
```

sftim commented on content/en/blog/_posts/2023-03-22-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,56 @@
+
+More detials can be found in the KEP <https://kep.k8s.io/1040> and <https://github.com/kubernetes/kubernetes/pull/116121>.
+
+## Evented PLEG will enhance the scalability of kubelet
```
```suggestion
## Event triggered updates to container status
```

sftim commented on content/en/blog/_posts/2023-03-22-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,56 @@
+
+To identify the cause of slow pod startup, analyzing metrics and logs can be helpful. Other factors that may impact pod startup include container runtime, disk speed, CPU and memory resources on the node.
+
+SIG Node is responsible for ensuring fast Pod startup times, while addressing issues in large clusters falls under the purview of SIG Scalability as well.
```
This doesn't seem to link to the previous text. Is it needed?

pacoxu replied:
I wanted users can know which SIGs care more about the pod startup speed.

sftim reviewed (commented):
Thanks. We'll need changes before we can publish this.

/hold
pending assignment of a publication date

tengqm commented on content/en/blog/_posts/2023-03-22-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,57 @@
+
+**Authors**: Paco Xu(DaoCloud), Sergey Kanzhelev(Google), Ruiwen Zhao(Google)
+
+How can pod start-up time be accelerated on nodes in large clusters? This is a common issue that cluster administrators may face.
```
Please manually wrap the long lines so that each line is about 100 characters or so.

Manual wrapping helps the reviewers to provide feedbacks to your text;
it will also help downstream localization teams to translate line by line and
for them to track future changes (git diff works by doing line-by-line comparison anyway).

pacoxu replied:
Thanks for the advice and the PR is updated.

pacoxu commented:
@SergeyKanzhelev 
I add another factor that is also changed in v1.27.
> 4. pod resource limit (most likely the CPU limit will influence it; BTW we changed the throttling factor for cgroup v2 memory, if pod uses a lot of memory during startup, this may be a possible factor)

pacoxu commented:
The SELinux Relabeling with Mount Options feature moved to Beta in v1.27. This feature speeds up container startup by mounting volumes with the correct SELInux label instead of changing each file on the volumes recursively.

https://github.com/kubernetes/kubernetes/pull/116425
@jsafrane I think this feature also helped speed up pod startup. 

I will add this in the end of the blog later.

pacoxu commented:
Updated.

tengqm commented on content/en/blog/_posts/2023-04-04-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+---
+layout: blog
+title: "Recent developments in kubelet to speed up Pod startup"
```
"Recent" is always relative to now.
As time goes by, the content won't stay afresh.
Better explicitly spell out the version you are referring to.

pacoxu replied:
Most updates are in v1.27 and only 1 mentioned below is v1.26.

Is `Developments updates in kubelet to speed up Pod startup`  better?

tengqm replied:
Maybe "1.27 updates on speeding up Pod startup" ?

tengqm commented on content/en/blog/_posts/2023-04-04-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+
+**Authors**: Paco Xu(DaoCloud), Sergey Kanzhelev(Google), Ruiwen Zhao(Google)
+
+How can pod start-up time be accelerated on nodes in large clusters? This is a common issue that
```
You accelerate Pod start-up, or your shorten the Pod start-up time ...

pacoxu replied:
Fixed.

tengqm commented on content/en/blog/_posts/2023-04-04-recent-developments-in-kubelet-to-speed-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+include scheduling time for pods or webhooks executed on it.
+
+We have mentioned some important factors here to consider from the kubelet's perspective, but
+this is not an exhaustive list. As Kubernetes v1.27 is set to release soon, this blog will highlight
+significant changes in v1.27 that aid in speeding up pod start-up.
```
> This blog **will highlight** ...

Something new is planned?

pacoxu replied:
Fixed.

tengqm commented:
LGTM from docs perspective.

pacoxu commented:
> /hold
> pending assignment of a publication date

I change the publication date to 2023-04-11 as the title changed to `1.27 updates on speeding up Pod startup`.

pacoxu commented:
> /hold
> pending assignment of a publication date

I changed to date to 04-14. Is this OK?

pacoxu commented:
/assign @ruiwen-zhao @SergeyKanzhelev @sftim

ruiwen-zhao commented:
/lgtm 

And since 1.27 is released now, this blog should be ready to be published. Thanks for all the effort Paco!

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+---
+layout: blog
+title: "1.27 updates on speeding up Pod startup"
```
```suggestion
title: "Kubernetes 1.27: updates on speeding up Pod startup"
```

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+slug: speed-up-pod-startup
+---
+
+**Authors**: Paco Xu(DaoCloud), Sergey Kanzhelev(Google), Ruiwen Zhao(Google)
```
```suggestion
**Authors**: Paco Xu (DaoCloud), Sergey Kanzhelev (Google), Ruiwen Zhao (Google)
```

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+involve the creation time of pods by controller-manager through kube-apiserver, nor does it
+include scheduling time for pods or webhooks executed on it.
+
+We have mentioned some important factors here to consider from the kubelet's perspective, but
+this is not an exhaustive list. As Kubernetes v1.27 is set to release soon, this blog highlights
+significant changes in v1.27 that aid in speeding up pod start-up.
```
You'll need to reword this now.

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+this is not an exhaustive list. As Kubernetes v1.27 is set to release soon, this blog highlights
+significant changes in v1.27 that aid in speeding up pod start-up.
+
+## Parallel Image Pulls
```
```suggestion
## Parallel container image pulls
```

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+
+### Maximum parallel image pulls will help secure your node from overloading on image pulling
+
+We are introducing a new feature in kubelet that sets a limit on the number of parallel image
+pulls at the node level. This limit restricts the maximum number of images that can be pulled
+simultaneously. If there is an image pull request beyond this limit, it will be blocked until
+one of the ongoing image pulls finishes. Before enabling this feature, please ensure that your
+container runtime's image service can handle parallel image pulls effectively.
```
You'll need to reword this now: the feature is released.

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+be pulled concurrently. Any additional image pulls beyond this limit will wait until at least
+one ongoing pull is complete.
+
+You can find more details in the associated KEP: [Kubelet limit of Parallel Image Pulls](https://kep.k8s.io/3673)
+ (KEP-3673). The improvement will arrive as part of Kubernetes v1.27 (scheduled for release in April 2023).
```
You'll need to reword this now; v1.27 is released. Ideally, also link to the docs for managing image pull parallelism.

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+secrets, or volumes. This requires a large bandwidth to access kube-apiserver.
+
+In versions prior to v1.27, the default `kubeAPIQPS` was 5 and `kubeAPIBurst` was 10. However,
+Kubelet has increased these defaults to 50 and 100 respectively for better performance during
```
```suggestion
the kubelet in v1.27 has increased these defaults to 50 and 100 respectively for better performance during
```

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+2. In large clusters they can generate significant load anyway as there are a lot of them
+3. They have a dedicated PriorityLevel and FlowSchema that we can easily control
+
+In old days, we always encounter `volume mount timeout` on kubelet in node with more than 50 pods
```
```suggestion
Previously, we often encountered `volume mount timeout` on kubelet in node with more than 50 pods
```
“the old days” refers to a time significantly further in the past (typically not this century).

sftim replied:
```suggestion
Previously, we often encountered `volume mount timeout` on kubelet in node with more than 50 pods
```
“the old days” refers to a time significantly further in the past (typically not this century).

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+3. They have a dedicated PriorityLevel and FlowSchema that we can easily control
+
+In old days, we always encounter `volume mount timeout` on kubelet in node with more than 50 pods
+during pod start up. We suggest our customers to bump `kubeAPIQPS` to 20 and `kubeAPIBurst` to 40
+and especially with customers that is using bare-metal(physical machines).
```
```suggestion
during pod start up. We suggest that cluster operators bump `kubeAPIQPS` to 20 and `kubeAPIBurst` to 40,
especially if using bare metal nodes.
```
I think this is better.

sftim replied:
```suggestion
during pod start up. We suggest that cluster operators bump `kubeAPIQPS` to 20 and `kubeAPIBurst` to 40,
especially if using bare metal nodes.
```
I think this is better.

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+
+## Event triggered updates to container status
+
+`Evented PLEG` (PLEG is short for "Pod Lifecycle Event Generator") is set to be in beta for v1.27,
```
```suggestion
Kubernetes offers two ways for the kubelet to detect Pod lifecycle events, such as a the last process in
a container shutting down. In Kubernetes v1.27, the _event based_ mechanism has graduated to beta but
remains
```

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+## Event triggered updates to container status
+
+`Evented PLEG` (PLEG is short for "Pod Lifecycle Event Generator") is set to be in beta for v1.27,
+disabled by default (needs to be enabled explicitly). It aids pod startup by reducing unnecessary
+work during inactivity through replacing periodic polling. The current `Generic PLEG` incurs
+non-negligible overhead due to frequent container status polling, which worsens with Kubelet's
+parallelism, leading to poor performance and reliability issues. For these reasons, it is
+recommended that we use Evented PLEG instead. 
```
```suggestion
disabled by default. If you do explicitly switch to event-based lifecycle change detection,
the kubelet is able to start Pods more quickly than with the default approach that relies on polling.
The default mechanism, polling for lifecycle changes, adds a noticeable overhead; this affects
the kubelet's ability to handle different tasks in parallel, and leads to poor performance and
reliability issues. For these reasons, we recommend that you switch your nodes to use
event-based pod lifecycle change detection.
```

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+parallelism, leading to poor performance and reliability issues. For these reasons, it is
+recommended that we use Evented PLEG instead. 
+
+Further details can be found in the KEP <https://kep.k8s.io/3386>.
```
Please link to the docs about pod lifecycle change detection (if there aren't any, write some and then link there).

pacoxu replied:
https://kubernetes.io/docs/tasks/administer-cluster/switch-to-evented-pleg/ link to this.

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+
+Further details can be found in the KEP <https://kep.k8s.io/3386>.
+
+## Raise your pod resource limit if needed
+
+During start-up, some pods may consume a considerable amount of CPU or memory. If the CPU limit is
+low, this can significantly slow down the pod start-up process. To improve the memory management,
+Kubernetes v1.22 introduced a feature gate called MemoryQoS to kubelet. This feature enables
+kubelet to set memory QoS at container, pod, and QoS levels for better protection and guaranteed
+quality of memory when running with cgroups v2. Although it has benefits, it is possible that
+enabling this feature gate may affect the start-up speed of the pod if the pod startup consumes
+a large amount of memory.
+
+Kubelet configuration now includes `memoryThrottlingFactor`. This factor is multiplied by
+the memory limit or node allocatable memory to set the cgroupv2 memory.high value for enforcing
+MemoryQoS. Decreasing this factor sets a lower high limit for container cgroups, increasing reclaim
+pressure. Increasing this factor will put less reclaim pressure. The default value is 0.8 initially
+and will change to 0.9 in Kubernetes v1.27. This parameter adjustment can reduce the potential
+impact of this feature on pod startup speed.
+
+Further details can be found in the KEP <https://kep.k8s.io/2570>.
```
I think https://github.com/kubernetes/website/pull/39853 covers this. I don't mind if we merge the two articles; that's even a good thing if we do. However, we need to manage how that happens and check if the other authors are OK with the idea.

pacoxu replied:
/cc @ndixita

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+> Dec 30 15:33:13.375379 e2e-022435249c-674b9-minion-group-gdj4 kubelet[8362]: I1230 15:33:13.375359    8362 pod_startup_latency_tracker.go:102] "Observed pod startup duration" pod="kube-system/konnectivity-agent-gnc9k" podStartSLOduration=-9.223372029479458e+09 pod.CreationTimestamp="2022-12-30 15:33:06 +0000 UTC" firstStartedPulling="2022-12-30 15:33:09.258791695 +0000 UTC m=+13.029631711" last ...
+
+The SELinux Relabeling with Mount Options feature moved to Beta in v1.27. This feature speeds up
+container startup by mounting volumes with the correct SELInux label instead of changing each file
```
```suggestion
container startup by mounting volumes with the correct SELinux label instead of changing each file
```

sftim replied:
```suggestion
container startup by mounting volumes with the correct SELinux label instead of changing each file
```

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+
+The SELinux Relabeling with Mount Options feature moved to Beta in v1.27. This feature speeds up
+container startup by mounting volumes with the correct SELInux label instead of changing each file
+on the volumes recursively. Further details can be found in the KEP <https://kep.k8s.io/1710>.
```
We should link to the blog article that https://github.com/kubernetes/website/pull/39836 adds (if it merges)

pacoxu replied:
Not sure about the process. 
Should this blog wait for that pr merge, or we can update it later after this blog is merged?

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+
+The SELinux Relabeling with Mount Options feature moved to Beta in v1.27. This feature speeds up
+container startup by mounting volumes with the correct SELInux label instead of changing each file
+on the volumes recursively. Further details can be found in the KEP <https://kep.k8s.io/1710>.
```
We should link to the blog article that https://github.com/kubernetes/website/pull/39836 adds (if it merges). The path will be `/blog/2023/04/18/kubernetes-1-27-efficient-selinux-relabeling-beta/`.

sftim commented on content/en/blog/_posts/2023-04-14-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,115 @@
+parallelism, leading to poor performance and reliability issues. For these reasons, it is
+recommended that we use Evented PLEG instead. 
+
+Further details can be found in the KEP <https://kep.k8s.io/3386>.
```
Please link to the docs about pod lifecycle change detection (if there aren't any, write some and then link there).

sftim reviewed (changes requested):
I don't think this article was one that the release comms team are tracking.

We should try to involve that team in planning articles that are linked to a release. Please say Hi in #release-comms on Kubernetes Slack and discuss a way forward. If you've already got in contact there, could you have one more go?

I like the ambition to cover more than one KEP in a single article; that's what readers will want. However, I can see that two KEPS covered here already have announcement articles signed off.

sftim commented:
/lgtm cancel

pacoxu commented:
> I don't think this article was one that the release comms team are tracking.
> 
> 
> 
> We should try to involve that team in planning articles that are linked to a release. Please say Hi in #release-comms on Kubernetes Slack and discuss a way forward. If you've already got in contact there, could you have one more go?
> 
> 
> 
> I like the ambition to cover more than one KEP in a single article; that's what readers will want. However, I can see that two KEPS covered here already have announcement articles signed off.

Thanks for the information. I will check with #release-commons

pacoxu commented:
> Hello, The publication date for this article is 15-05-2023 (May 15).
> 
> 
> 
> Thank you!

Nice. Then I can follow it asap after the kubecon.

SergeyKanzhelev reviewed (approved):
/lgtm

pacoxu commented:
> Hello, The publication date for this article is 15-05-2023 (May 15).

The date is updated.

stmcginnis commented:
/lgtm

tengqm commented on content/en/blog/_posts/2023-05-15-kubernetes-1.27-updates-on-speeding-up-pod-startup.md:
```diff
@@ -0,0 +1,120 @@
+---
+layout: blog
+title: "Kubernetes 1.27: updates on speeding up Pod startup"
```
Better avoid dots in the file name, replace `1.27` with `1-27`.

pacoxu replied:
Updated. Thanks.

stmcginnis commented:
/lgtm

tengqm commented:
/hold cancel
/approve
